runtime-sdk/src/dispatcher.rs

```diff
@@ -467,8 +447,148 @@ impl<R: Runtime + Send + Sync> transaction::dispatcher::Dispatcher for Dispatche
+        let mut tx_reject_hashes = Vec::new();
+
+        let mut result = self.execute_batch_common(
+            rt_ctx,
+            |ctx| -> Result<Vec<ExecuteTxResult>, RuntimeError> {
+                // Schedule and execute the batch.
+                //
+                // The idea is to keep scheduling transactions as long as we have some space
+                // available in the block as determined by gas use.
+                let mut new_batch = Vec::new();
+                let mut results = Vec::with_capacity(batch.len());
+                let mut requested_batch_len = cfg.initial_batch_size;
+                'batch: loop {
+                    // Remember length of last batch.
+                    let last_batch_len = batch.len();
+                    let last_batch_tx_hash = batch.last().map(|raw_tx| Hash::digest_bytes(raw_tx));
+
+                    for raw_tx in batch.drain(..) {
+                        // If we don't have enough gas for processing even the cheapest transaction
+                        // we are done.
+                        let remaining_gas = R::Core::remaining_batch_gas(ctx);
+                        if remaining_gas < cfg.min_remaining_gas {
+                            break 'batch;
+                        }
+
+                        // Decode transaction.
+                        let tx = match Self::decode_tx(ctx, &raw_tx) {
+                            Ok(tx) => tx,
+                            Err(_) => {
+                                // Transaction is malformed, make sure it gets removed from the
+                                // queue and don't include it in a block.
+                                tx_reject_hashes.push(Hash::digest_bytes(&raw_tx));
+                                continue;
+                            }
+                        };
+                        let tx_size = raw_tx.len().try_into().unwrap();
+
+                        // If we don't have enough gas remaining to process this transaction, just
+                        // skip it.
+                        if tx.auth_info.fee.gas > remaining_gas {
+                            continue;
+                        }
+                        // Same if we don't have enough consensus message slots.
+                        if tx.auth_info.fee.consensus_messages > ctx.remaining_messages() {
+                            continue;
+                        }
+
+                        // Determine the current transaction index.
+                        let index = new_batch.len();
+
+                        new_batch.push(raw_tx);
+                        results.push(Self::execute_tx(ctx, tx_size, tx, index)?);
+                    }
+
+                    // If there's more room in the block and we got the maximum number of
+                    // transactions, request more transactions.
+                    if last_batch_tx_hash.is_some()
```

Comment:
I'd personally make this function operation on an iterator of batches, but that's just because I can't think about tracking so many little bits of state :P